### PR TITLE
feat: log local TCP address

### PR DESCRIPTION
**Pull request by maxmoehl**
Summary
---------------

When troubleshooting issues related to keepalive it is often useful to be able to correlate multiple requests made over the same connection. A first step was to introduce the information on whether the request was made on a re-used connection or not which was done with a7dbf849. This commit adds the local address of the TCP connection which adds the capability to distinguish between multiple, parallel keep-alive connections by looking at the client port of gorouter for the backend connection.

- [x] Read the [Contributing document](../blob/-/.github/CONTRIBUTING.md).

Backward Compatibility
---------------
Breaking Change? **Yes**

### CODE REVIEW

**Comment by maxmoehl** [comment]
The big question is: how do we proceed with such log changes which could be considered breaking? I don't think it's a healthy practice to keep adding flags for each and every individual log field we add 🤔

**Comment by maxmoehl** [comment]
> sadly this is a breaking change (whomp)

That's what I was expecting 😄 

> Develop a more advanced config property that allows users to "opt in" to certain fields by name 1 by 1. So that way we can add new fields, but they can always opt in / out of certain ones. - Fixes this problem for the future. But this is more work. And this is just a config property for everything! Will operators even opt in to any of them!?

That's probably what I will be going for in this PR to prevent accumulating further config flags. Just a flat list, and if anyone needs special properties they can be added and exposed via this mechanism without breaking anyone.

> add an additional access-log file that's json-based for people to migrate to, and we deprecate the old one eventually. When it is json based we can add new properties without it being a breaking change. - Personally I love this idea, but it is definitely more work.

This is my preferred long-term solution, but it will require quite some effort so it's not clear when we will be able to invest.

File: accesslog/schema/access_log_record_test.go [hunk 1]
```diff
@@ -314,6 +314,27 @@ var _ = Describe("AccessLogRecord", func() {
 				Eventually(r).Should(Say(`x_cf_routererror:"-"`))
 			})
 		})
+
+		Context("when extra_fields is set to [local_address]", func() {
```
Comment (ameowlia) [inline accesslog/schema/access_log_record_test.go:318]: Can you add an test for...
* when extra_fields does not contain local_address, then the local address is not tested.
* when extra_fields contains "meow" it ... does not error? logs? idk does something reasonable?

**Review by ameowlia — CHANGES_REQUESTED** [review]
❓ Can you add a PR to routing-release so I can deploy and test everything at once?

**Comment by maxmoehl** [comment]
I've opened cloudfoundry/routing-release#448 as a draft as I'm still adding tests.

File: accesslog/schema/access_log_record_test.go (hunk 1 above)
Comment (maxmoehl) [reply accesslog/schema/access_log_record_test.go:318]: I've added two more test cases.

**Comment by maxmoehl** [comment]
The past weeks have been a bit busy, I finally found the time to clean this up. The [rr PR](https://github.com/cloudfoundry/routing-release/pull/448) is also ready now. Unfortunately test execution on my end is still hindered by cloudfoundry/routing-release#450. I ran the package tests using `go test` and they seem fine but I can't run the whole shebang, could you please do that? Thank you!

I'm currently also deploying this to one of our environments.

File: accesslog/schema/access_log_record.go [hunk 2]
```diff
@@ -319,6 +322,17 @@ func (r *AccessLogRecord) makeRecord(performTruncate bool) []byte {
 		b.WriteDashOrFloatValue(r.successfulAttemptTime())
 	}
 
+	// We have to consider the impact of iterating over a list. This technically allows to repeat
+	// some of the fields but it allows us to iterate over the list only once instead of once per
+	// field when we perform a [slices.Contains] check.
+	for _, field := range r.ExtraFields {
```
Comment (ameowlia) [inline accesslog/schema/access_log_record.go:328]: If we iterate over the list this means the operator will get to determine the order of these, vs a standardized order. 

Does that matter?

File: proxy/round_tripper/proxy_round_tripper.go [hunk 3]
```diff
@@ -191,6 +191,7 @@ func (rt *roundTripper) RoundTrip(originalRequest *http.Request) (*http.Response
 				slog.Float64("dns-lookup-time", trace.DnsTime()),
 				slog.Float64("dial-time", trace.DialTime()),
 				slog.Float64("tls-handshake-time", trace.TlsTime()),
+				slog.String("local-addr", trace.LocalAddr()),
```
Comment (ameowlia) [inline proxy/round_tripper/proxy_round_tripper.go:194]: ❓ Why call it "local-addr" here, but "local_address" above? Why not "local-address" here?

File: proxy/round_tripper/proxy_round_tripper.go [hunk 4]
```diff
@@ -255,6 +256,7 @@ func (rt *roundTripper) RoundTrip(originalRequest *http.Request) (*http.Response
...
```
Comment (ameowlia) [inline proxy/round_tripper/proxy_round_tripper.go:259]: From reading your spec description on the other PR, it is not clear if the "extra_headers" property is only for access logs or not. 

❓ Based on my reading of it, I would've thought it would affect these logs too? Thoughts?

**Comment by ameowlia** [comment]
✅ tests pass

File: proxy/round_tripper/proxy_round_tripper.go (hunk 4 above)
Comment (maxmoehl) [reply proxy/round_tripper/proxy_round_tripper.go:259]: Hm, in the past we didn't give much thought to the compatibility of non-access-logs, primarily because these are JSON and we expect people to parse it as such. Additional fields shouldn't cause any breakage in those cases. For example the timing info directly above this line is controlled by a feature flag in the access logs but unconditionally included in this log.

File: accesslog/schema/access_log_record.go (hunk 2 above)
Comment (maxmoehl) [reply accesslog/schema/access_log_record.go:328]: It probably does matter. The other thought I had is that you can end up with multiple instances of the same field (maybe you use some complex templating for the manifest and accidentally add it twice that way). What about this: during the creation of the access logger we de-duplicate the list and order it alphabetically. That way we have a consistent ordering and prevent people from having duplicate fields.

The only concern would then be if someone intentionally wants to influence the order for whatever reason but we usually don't allow that anyway.

File: proxy/round_tripper/proxy_round_tripper.go (hunk 3 above)
Comment (maxmoehl) [reply proxy/round_tripper/proxy_round_tripper.go]: Fixed it.

File: accesslog/schema/access_log_record.go (hunk 2 above)
Comment (ameowlia) [reply accesslog/schema/access_log_record.go:328]: > What about this: during the creation of the access logger we de-duplicate the list and order it alphabetically. That way we have a consistent ordering and prevent people from having duplicate fields.

I like the idea of de-duplicating.

I theoretically like the idea of alphabetizing, but I don't like that I think it would cause breaking changes. I know we only have one "extra field" here right now. But I think eventually we want to put all (most??) access logger fields to be controlled by this property, correct? And if it alphabetizes it, then the order would change.

tl;dr:
✅ deduping
⛔ alphabetizing imo

File: proxy/round_tripper/proxy_round_tripper.go (hunk 4 above)
Comment (ameowlia) [reply proxy/round_tripper/proxy_round_tripper.go:259]: > I could adjust the wording in the release spec to reflect that clearly.

Sounds good to me.

File: accesslog/schema/access_log_record.go (hunk 2 above)
Comment (maxmoehl) [reply accesslog/schema/access_log_record.go:328]: Fields are now de-duplicated!

**Review by ameowlia — APPROVED** [review]
